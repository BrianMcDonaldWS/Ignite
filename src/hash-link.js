import path from 'path';
import root from 'root-path';
import transformLinks from 'transform-markdown-links';

export default function(source) {
  const markdown = transformLinks(source, link => {
    if (path.extname(link) === '.md') {
      return `#${link}`;
    }

    const [filePath, ...description] = link.split(' ');
    const pathToThisSource = path.dirname(path.resolve(this.resourcePath));
    const pathToLink = path.join(pathToThisSource, filePath);
    const correctPath = path.relative(root(), pathToLink);

    return `${correctPath} ${description.join(' ')}`;
  });

  return markdown;
}
