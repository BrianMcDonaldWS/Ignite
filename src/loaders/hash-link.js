import path from 'path';
import transformLinks from 'transform-markdown-links';
import { getOptions } from 'loader-utils';

export function transformLink(resourcePath, link, options) {
  if (!link || link === '' || link.includes('http')) {
    return link;
  }

  const pathToThisSource = path.dirname(path.resolve(resourcePath));

  if (link[0] === '#') {
    return link;
  }

  if (path.extname(link).includes('.md')) {
    const pathToLink = path.join(pathToThisSource, link);
    const pathToDocs = path.join(process.cwd(), options.src);
    const correctPath = path.relative(pathToDocs, pathToLink);

    return path
      .join(
        options.static && !options.watch ? path.join(options.static, '/') : '/',
        correctPath
      )
      .replace('.md', '.html');
  }

  const [filePath, ...description] = link.split(' ');
  const pathToLink = path.join(pathToThisSource, filePath);
  const correctPath = path.relative(process.cwd(), pathToLink);

  return `${correctPath} ${description.join(' ')}`;
}

export function transform(source, resourcePath, options) {
  const markdown = transformLinks(source, link =>
    transformLink(resourcePath, link, options)
  );

  return markdown;
}

export default function(source) {
  const options = getOptions(this);
  return transform(source, this.resourcePath, options);
}
