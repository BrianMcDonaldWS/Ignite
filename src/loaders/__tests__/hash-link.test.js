import { transformLink, transform } from '../hash-link';

describe('transformLink', () => {
  beforeAll(() => {
    jest.spyOn(process, 'cwd').mockReturnValue('/project');
  });

  afterAll(() => {
    process.cwd.mockRestore();
  });

  test('should leave id links alone', () => {
    expect(transformLink('/path/', '#link')).toBe('#link');
  });

  test('should leave http links alone', () => {
    expect(
      transformLink(
        '/path/',
        'https://images.pexels.com/photos/20787/pexels-photo.jpg?auto=compress&cs=tinysrgb&h=350'
      )
    ).toBe(
      'https://images.pexels.com/photos/20787/pexels-photo.jpg?auto=compress&cs=tinysrgb&h=350'
    );
  });

  test('should hash link', () => {
    expect(
      transformLink('/docs/example.md', './markdown.md', { src: 'docs/' })
    ).toBe('../docs/markdown.md');
  });

  test('should make relative paths to images', () => {
    expect(
      transformLink('/path/', './link/to/image.png', { src: 'docs/' })
    ).toBe('../link/to/image.png ');
    expect(
      transformLink('/path/', './link/to/image.png "With a description"', {
        src: 'docs/'
      })
    ).toBe('../link/to/image.png "With a description"');
  });
});

describe('transform', () => {
  const options = {
    src: 'docs/'
  };

  expect(
    transform(
      `
    [Link](#foo)
    [Another](pages/intro.md)
  `,
      '/path',
      options
    )
  ).toMatchSnapshot();
});
