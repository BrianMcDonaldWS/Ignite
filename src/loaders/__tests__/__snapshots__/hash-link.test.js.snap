// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"
    [Link](#foo)
    [Another](#/pages/intro.md)
  "
`;
